lib/jekyll/reader.rb

```diff
@@ -16,9 +16,19 @@ def read
       read_directories
       read_included_excludes
       sort_files!
-      @site.data = DataReader.new(site).read(site.config["data_dir"])
       CollectionReader.new(site).read
       ThemeAssetsReader.new(site).read
+      @site.data = read_data
+    end
+
+    # Read and merge the contents of _data folders. Data from Site takes
+    # precedence over data from theme.
+    #
+    # Returns merged data hash.
+    def read_data
+      site_data = DataReader.new(site).read(site.config["data_dir"])
+      theme_data = DataReader.new(site).read(site.theme&.data_path)
+      Jekyll::Utils.deep_merge_hashes(theme_data, site_data)
```

Comment:
With our change to setting `@site.data` here, let's make the theme a bit more conditional. I'll read directly into `@site.data`, then guard the remainder of the method with a check to see if the theme exists.

```suggestion
      @site.data = DataReader.new(site).read(site.config["data_dir"])
      return unless site.theme&.data_path
      
      theme_data = DataReader.new(site, in_source_dir: site.method(:in_theme_dir)).read(site.theme&.data_path)
      @site.data = Jekyll::Utils.deep_merge_hashes(theme_data, @site.data)
```

This will require some changes to the `DataReader` class as well.

```diff
diff --git lib/jekyll/readers/data_reader.rb lib/jekyll/readers/data_reader.rb
index 6fbc047fa..8e81254aa 100644
--- lib/jekyll/readers/data_reader.rb
+++ lib/jekyll/readers/data_reader.rb
@@ -4,11 +4,12 @@ module Jekyll
   class DataReader
     attr_reader :site, :content
 
-    def initialize(site)
+    def initialize(site, in_source_dir: nil)
       @site = site
       @content = {}
       @entry_filter = EntryFilter.new(site)
-      @source_dir = site.in_source_dir("/")
+      @in_source_dir = in_source_dir || @site.method(:in_source_dir)
+      @source_dir = in_source_dir.call("/")
     end
 
     # Read all the files in <dir> and adds them to @content
@@ -18,7 +19,7 @@ module Jekyll
     # Returns @content, a Hash of the .yaml, .yml,
     # .json, and .csv files in the base directory
     def read(dir)
-      base = site.in_source_dir(dir)
+      base = @in_source_dir.call(dir)
       read_data_to(base, @content)
       @content
     end
@@ -38,7 +39,7 @@ module Jekyll
       end
 
       entries.each do |entry|
-        path = @site.in_source_dir(dir, entry)
+        path = @in_source_dir(dir, entry)
         next if @entry_filter.symlink?(path)
 
         if File.directory?(path)
```